File: backend/schemas/importacion.py
"""
Schemas Pydantic para el módulo de importación masiva de NÓMINA de sueldos (costos_nomina).
El import de empleados (roster) vive en schemas/importacion_nomina_empleados.py.
"""
from typing import List

from pydantic import BaseModel


# ─── Comunes ─────────────────────────────────────────────────────────────────

class FilaError(BaseModel):
    fila: int
    campo: str
    error: str


class ConfirmarError(BaseModel):
    fila: int
    error: str


# ─── Nómina (sueldos → costos_nomina) ────────────────────────────────────────

class FilaNominaPreview(BaseModel):
    fila: int
    dni: str
    nombre_empleado: str   # resuelto via DNI→empleado en el preview
    empleado_id: str       # UUID del empleado, necesario para el confirmar
    anio: int
    mes: int
    salario_bruto: float
    neto: float
    es_actualizacion: bool = False  # True si ya existe nómina para (empleado_id, anio, mes)


class ImportacionNominaPreviewResponse(BaseModel):
    filas_validas: List[FilaNominaPreview]
    errores: List[FilaError]


class ImportacionNominaConfirmarRequest(BaseModel):
    empresa_id: str  # solo para trazabilidad; empresa_id real se hereda del empleado en el repo
    filas: List[FilaNominaPreview]


class ImportacionNominaConfirmarResponse(BaseModel):
    importados: int
    actualizados: int
    errores: List[ConfirmarError]

File: frontend/types/importacion.ts
// ─── Comunes ────────────────────────────────────────────────────────────────

export interface FilaError {
  fila: number
  campo: string
  error: string
}

export interface ConfirmarError {
  fila: number
  error: string
}

// ─── Nómina de empleados (roster, 27 columnas) ──────────────────────────────

export interface FilaConFaltantes {
  fila: number
  empleado: string
  faltan: string[]
}

export interface FilaNoCargada {
  fila: number
  empleado: string
  motivo: string
}

export interface ImportacionNominaEmpleadosResult {
  total: number
  creados: number       // altas nuevas (DNI no existía)
  actualizados: number  // updates (DNI ya existía) — dedup
  cargados_ok: number   // cargados sin faltantes
  con_faltantes: FilaConFaltantes[]
  no_cargados: FilaNoCargada[]
}

// ─── Nómina de sueldos (costos_nomina) ──────────────────────────────────────

export interface FilaNominaPreview {
  fila: number
  dni: string
  nombre_empleado: string
  empleado_id: string
  anio: number
  mes: number
  salario_bruto: number
  neto: number
  es_actualizacion: boolean
}

export interface ImportacionNominaPreview {
  filas_validas: FilaNominaPreview[]
  errores: FilaError[]
}

export interface ImportacionNominaResult {
  importados: number
  actualizados: number
  errores: ConfirmarError[]
}
